### ISSUE TITLE
What should we call the anomaly type/category field?

### ISSUE THREAD

**Issue by hellais** [TheTorProject/ooni-measurements#32]
> There are only 2 hard things in computer science ...  bla bla bla

We have come up with some good categories that are the following:

This is used to indicate the type of the anomaly.
The possible anomaly types are:
    * `confirmed`: for confirmed network anomalies (we found a blockpage, a middlebox was found, the IM app is blocked, etc.)
    * `anomaly`: this measurement requires special attention (it's likely to be a case of blocking), however it has not been confirmed
    * `failure`: there was an error in the measurement (the control request failed, there was a bug, etc.)
    * `ok`: we are "sure" there was no network anomaly

This is going to be returned inside of the `/api/v1/measurements` endpoint and users will be able to filter by it.

Now the $1M question is: how do we call this field key?

Some of the suggestions that came up are:

* `anomaly_level`
* `measurement_result_type`
* `discovery`
* `measurement_diagnosis`
* `measurement_verdict`
* `measurement_score`
* `measurement_product`
* `measurement_observation`
* `measurement_answer`
* `measurement_outcome`
* `measurement_bikeshed_colour`
* `measurement_bikeshed_color`

Let the bikeshedding begin!

**Comment by bassosimone** [comment]
Meta bikeshedding: should anyone expressed a preference, should we use the Schultze method? What should we do if anything new is proposed? :trollface:

**Comment by bassosimone** [comment]
Thinking a bit more, if the objective is to determine the `type of anomaly`, I'd say:

```JSON
anomaly_type = [
    "confirmed",
    "possible",
    "measurement_failed",
    "none"
]
```

**Comment by darkk** [comment]
IMHO **type** of anomaly is something like:
- tcp/ip filtering
- IP address blackholing
- transparent proxy
- RST injection
- DNS poisioning/injection/spoofing
- content injection (e.g. <iframe/> with some ads from ISP)
- blockpage
- wrong Date header
- ...

And it's incomplete list of anomaly types. On the other hand set of _verdicts_ w're discussing in this ticket is set once and forever.

**Comment by hellais** [comment]
> And it's incomplete list of anomaly types. On the other hand set of verdicts w're discussing in this ticket is set once and forever.

While I agree that there may be some confusion with calling it type, I don't think what we are talking about here is going to be set once and forever. For example something can become `confirmed` once a human has classified it as such.

Maybe it's more appropriate to call it `measurement_result_class`?

**Comment by hellais** [comment]
>> use the Schultze method

> There is app for that. https://modernballots.com/

I am not sure voting on something like this is the best way to go since it's actually something that determines the language we use in communicating about our results to some extent.

Maybe asking people from the community what they think about it

**Comment by hellais** [comment]
> "measurement_failed"

I don't think `measurement_failed` is that accurate as `failure` could also be associated with a bug (also not related to the measurement itself) or a control failure.
Though I don't have a strong opinion on this.

> "none"
I would avoid using this as it can create confusion with python `None` which is serialised in JSON as `"key": null`

**Comment by hellais** [comment]
I am going to make a proposal that maybe we can get some consensus on.

How about instead of exposing an `enum` we expose multiple binary keys to indicate different states.

For example:

* `failure`: True or False
* `confirmed`: True or False
* `anomaly`: True or False

The state OK would basically just be determined from whether or not `anomaly & confirmed & failure` are all false.

This separation also allows us to express some more cases such as: there was a failure in the measurement (because the control failed), but actually the `confirmed` state is `True` because we saw a known blockpage.

What do you think? 👍 👎 ?

**Comment by agrabeli** [comment]
Sounds good, makes sense.

**Comment by hellais** [comment]
On the topic of API design this is a good read: http://zalando.github.io/restful-api-guidelines.

I am glad to see I am actually following most of their recommendations in ooni-measurements.

**Comment by bassosimone** [comment]
I don't think this is a good attempt at solving the problem we had (most often, when you reach the level of bikeshedding, you're discussing a quasi non-problem). What is interesting is the part where you show that we can have a combination of results: i.e., blockpage and failed control. This is brilliant in that the discussion is now not focused anymore on a name, but on what we can express. 👍

**Comment by hellais** [comment]
> or null... What does false mean? Is it confirmed negative or not confirmed positive ?

I would avoid using `null` to define boolean keys as a boolean is actually just a binary value (in this case it would become a tertiary value).

I would say `False` simply means that our heuristics for confirming blockage (currently only blockpage fingerprints) didn't produce a positive result.

If you care to take into account also the nuance of "confirmed negative" then we should introduce another boolean key that is called something like "accessible" or "normal". Do we even have heuristics for this at the moment?

My preference for encoding this sort of information, though, would actually be to set all the keys to `false` (anomaly = false, confirmed = false, failure = false).

**Comment by hellais** [comment]
Alternatively if you think it's important to use confirmed for encoding this information I would be OK with exposing something like:

```
confirmed:
  type: string
  enum:
     - yes # we confirmed the blocking
     - no # we confirmed the non-blocking
     - unknown # we cannot confirm nor deny
```
See: http://zalando.github.io/restful-api-guidelines/#122

**Comment by hellais** [comment]
> true is known failure, false is everything parsed and no failure indicators found, what should it be if pipeline sees bunch of unknown fields in json? We don't have our data strongly typed. Should it be null?

I would classify a bunch of unknown field in JSON as `failure: true` as well. This is a sign of some failure in the measurement that may be caused by some misbehaving client or a bug.

The trustworthiness of the measurement should therefore be considered less.

**Comment by kaerumy** [comment]
Trying to understand, why the medium.com DNS tests for Malaysia would always be marked:

anomaly: false
failure: true

https://api.ooni.io/api/v1/measurements?probe_cc=MY&input=medium.com

From reading this thread, that it would mean that there is a problem with the measurements, and that it cannot be used reliably? Then all of our measurements are bad? 

Reading the results, it's clear for me that there is DNS blocking. 

OONI Explorer for example marks it these results as signs of tampering, reason dns.

https://explorer.ooni.torproject.org/measurement/20171001T022635Z_AS17971_jxGt33hgoscUTn4qEOp2lRI2tvTtwDKSbJIZGt9p5EP07zkau5?input=https:%2F%2Fmedium.com

**Comment by hellais** [comment]
> From reading this thread, that it would mean that there is a problem with the measurements, and that it cannot be used reliably? Then all of our measurements are bad and cannot be used according to the API results?

@darkk should correct me if I am wrong, but I believe that we currently don't take into account DNS heuristics in the pipeline when annotating measurements with the `confirmed` flag. The lack of the `anomaly` flag is unusual, though, and they are marked as `failure` likely to an issue in processing them at the pipeline stage.

We will look into this and figure out why they are being flagged as failing (`failure` means that there was some issue in processing the data, possibly due to it being malformed).

**Comment by kaerumy** [comment]
Thanks. In theory it should be confirmed, because at least for Medium.com it should be getting the block page with the alternative dns provided by the ISP  https://github.com/TheTorProject/ooni-pipeline/issues/35

**Comment by Jeffrey04** [comment]
Just need some clarification on 

```The state OK would basically just be determined from whether or not anomaly & confirmed & failure are all false.```

should that be either one of `anomaly` `confirmed` `failure` is `false`, or they have to be all `false` to indicate the site is unreachable?

**Comment by Jeffrey04** [comment]
```
The state OK would basically just be determined from whether or not anomaly & confirmed & failure are all false.
```

I am building a dashboard to display a summary for the measurements. However if I build the dashboard according to the statement I would have all the measurements are treated as failure to access, because 

```
mysql> select * from measurements where failure = false limit 10;
Empty set (0.78 sec)
```

None of the measurement results returned by the API is failure free.

**Comment by hellais** [comment]
> None of the measurement results returned by the API is failure = false

Yes this is currently a bug in the data processing pipeline.

See: https://github.com/ooni/api/issues/59

**Comment by Jeffrey04** [comment]
so ideally the original statement is true (site is up when all the flags are `false`), at least after bug #59 is fixed? Meanwhile I would have to treat the measurements as "normal" because `anomaly` and `confirmed` are both `false`.
